dali/util/numpy.cc

```diff
@@ -183,45 +183,51 @@ void ParseHeaderItself(HeaderData &parsed_header, char *data, size_t header_len)
 
 void ParseODirectHeader(HeaderData &parsed_header, InputStream *src, size_t o_direct_alignm,
                         size_t o_direct_read_len_alignm) {
-  size_t aligned_len = align_up(10, o_direct_read_len_alignm);
-  auto token_mem = mm::alloc_raw_shared<char, mm::memory_kind::host>(aligned_len, o_direct_alignm);
+  const size_t token_len = 6 + 1 + 1 + 2;
+  size_t token_read_len = align_up(token_len + 1, o_direct_read_len_alignm);
+  auto token_mem =
+      mm::alloc_raw_shared<char, mm::memory_kind::host>(token_read_len, o_direct_alignm);
   char *token = token_mem.get();
-  auto file = dynamic_cast<ODirectFileStream*>(src);
-  int64_t nread = file->ReadAt(token, aligned_len, 0);
-  DALI_ENFORCE(nread <= static_cast<Index>(aligned_len) &&
-               nread >= static_cast<Index>(std::min(src->Size(), aligned_len)),
+  auto file = dynamic_cast<ODirectFileStream *>(src);
+  DALI_ENFORCE(
+      file,
+      "Could not read the numpy file header: expected file stream opened with O_DIRECT flag.");
+  int64_t nread = file->ReadAt(token, token_read_len, 0);
+  DALI_ENFORCE(nread <= static_cast<Index>(token_read_len) &&
+                   nread >= static_cast<Index>(std::min(src->Size(), token_read_len)),
                make_string("Can not read header: ",
-                           static_cast<Index>(std::min(src->Size(), aligned_len)),
-                           " <= ", nread, " <= ", aligned_len));
-  auto char_tmp = token[10];
-  token[10] = '\0';
+                           static_cast<Index>(std::min(src->Size(), token_read_len)), " <= ", nread,
+                           " <= ", token_read_len));
+  auto char_tmp = token[token_len];
+  token[token_len] = '\0';
 
   auto header_len = GetHeaderLen(token);
 
-  int64 offset = 6 + 1 + 1 + 2;
-  // the header_len can be 4GiB according to the NPYv2 file format
-  // specification: https://numpy.org/neps/nep-0001-npy-format.html
-  // while this allocation could be sizable, it is performed on the host.
-  auto block_start = align_down(offset, o_direct_alignm);
-  auto block_end = align_up(header_len + 1, o_direct_alignm);
-  aligned_len = align_up(block_end - block_start, o_direct_read_len_alignm);
+  // The header_len have up to 2**16 - 1 bytes. We do not support V2 headers
```

Comment:
Shouldn't we enforce V1 in such a case? Or V2 can be also a plain numpy file without structured arrays.

Reply:
So how about adding a warning that if v2 is encountered some features may not work.